string.c

```diff
@@ -2927,6 +2927,22 @@ rb_str_append(VALUE str, VALUE str2)
     return rb_str_buf_append(str, str2);
 }
 
+
+VALUE
+rb_str_append2(VALUE str, VALUE str2)
+{
+    StringValue(str2);
+    
+    if (RB_ENCODING_IS_ASCII8BIT(str) || ENCODING_GET(str) == ENCODING_GET(str2)) {
```

Comment:
That check is too simple, it should be something like "is str encoding's a superset of str2's encoding?".
e.g. it's fine to append US-ASCII to UTF-8.

Reply:
Or alternatively make this method only work on binary strings and name it `byteappend` (see issue).